collections/_article/d3d12-shaders-with-a-twist.md

```diff
@@ -0,0 +1,112 @@
+
+That worked quite well, but wasn't without big headaches about writing the code that should match the resource bindings in the Vulkan-targeted shaders to those in the DXIL ones. I won't delve into details on that here. Just know it required an insane amount of code and relying on certain SPIRV-Cross specifics. For now, the takeaway is that it was a hell of an approach, but worked in practice! (If you're curious, in the [first version of the pull request](https://github.com/godotengine/godot/pull/64304), you can see for yourself what I'm talking about.) This process also allowed me to make a few little contributions to SPIRV-Cross itself and learn its internals, which is always a good thing.
+
+### Second round: Via Mesa's NIR
+
+Months went by and I verified that the Microsoft was making great progress with **Dozen**, the Vulkan-to-D3D12 translation layer found in **Mesa**.
+
+Mesa was already capable of translating SPIR-V into its own intermediate representation (**NIR**), but Dozen would have added the ability to translate NIR to DXIL. That looked very promising because it would be a binary-to-binary translation (i.e., SPIR-V to DXIL) without having to go all the way back to a high level language (GLSL/HLSL), which is slower and less convenient.
+
+With some experimentation, I could make a shell script (included in the source code of the D3D12 renderer) that extracts from the Mesa source tree the relevant pieces of code to make the SPIR-V to DXIL translation possible.
+
+I wouldn't say that this integration was a piece of cake, but had major benefits in comparison to the SPIRV-Cross based approach. For instance, the bindings matching issue was easily solved by a Godot-specific patch to the Mesa code by which the translation process gets some callbacks. With that mechanism, the D3D12 renderer is told about the original set-binding pair of every resource in the SPIR-V shader and then it trivially maps them to some register indices in the DXIL world with some simple arithmetic:
+
+> DXIL register = SPIR-V set * 100000000 + binding * 100000
+
+A couple of notes on this:
+- We can make such a liberal use of DXIL registers since they are logical, just a way of tagging the bindings. (In past versions that was not the case, as far as I can tell.)
+- Regarding the last multiplication, if we understand Vulkan numbered bindings as "slots" and so we do with Direct3D register indices, in principle there would be no reason for it. However, Vulkan slots can hold entire arrays, whereas D3D ones need one slot per array element. Therefore, with that multiplication we are reserving a big range of D3D bindings to account for however big a incoming Vulkan array may be.
```

Comment:
```suggestion
A couple of notes on this:

- We can make such a liberal use of DXIL registers since they are logical, just a way of tagging the bindings. (In past versions that was not the case, as far as I can tell.)
- Regarding the last multiplication, if we understand Vulkan numbered bindings as "slots" and so we do with Direct3D register indices, in principle there would be no reason for it. However, Vulkan slots can hold entire arrays, whereas D3D ones need one slot per array element. Therefore, with that multiplication, we are reserving a big range of D3D bindings to account for however big an incoming Vulkan array may be.
```